src/dune_file.ml

```diff
@@ -1924,12 +1924,15 @@ module Documentation = struct
 end
 
 module Include_subdirs = struct
-  type t = No | Unqualified
+
+  type qualification = Unqualified | Qualified
+  type t = No | Include of qualification
 
   let decode =
     enum
       [ "no", No
-      ; "unqualified", Unqualified
+      ; "unqualified", Include Unqualified
+      ; "qualified", Include Qualified
```

Comment:
if it's not supported, why don't we fail at parsing time here?

Reply:
It is supported by the Coq mode only.

Reply:
That's why I submitted a PR separately, to introduce partial support for `qualified` before merging the Coq PR.

I got really bad feedback on the Coq mode using `unqualified` for `qualified`

Reply:
I don't see a better way to do this given that `include_subdirs` scope is multi-stanza; but ideas are welcome.

Reply:
How about making it parseable only if the coq extension is enabled in the dune-project?

Reply:
Sounds good, will do like this; however the check in `dir_contents` is still needed unfortunately :(

Reply:
If it's needed then it's fine to keep it in. The key is that we should error as soon as possible.

Why is it still needed btw?

Reply:
> Why is it still needed btw?

Imagine a project with `(using coq)` enabled, then in a dir we have a `dune` file:
```
(library (name foo) ...)
(coqlib (name Bar) ...)
(include_subdirs qualified)
```
we still need to guard the processing for the `foo` OCaml lib.

Reply:
I see. Well that's fine with me

Reply:
Cool, thanks!

Can I merge then? @emillon @diml @aalekseyev any other comments / concerns?

Reply:
A few random thoughts:
- I don't understand why we unconditionally fail in the `qualified` case. Shouldn't we only fail if there are some OCaml modules defined? (I presume this changes in the Coq PR, but I think logically it makes more sense here)
- Why does `c_sources` need the check? We're not making the C symbols qualified, are we? (they aren't even qualified by library, I think)
- I guess we could further allow `qualified` if the only .ml modules are in the top-level directory (because it doesn't make any difference there)

And separately, to make "local" cheap we could just reverse it. (although I have to say this is probably a trivial cost for anything but very very deep groups)

I'm happy if this is merged and any follow-up changes happen later.

Reply:
Thanks a lot @aalekseyev for the comments!

> * I don't understand why we unconditionally fail in the `qualified` case. Shouldn't we only fail if there are some OCaml modules defined?

I thought that would be safer in the sense that to determine if some OCaml modules are defined for the stanza you already have to interpret the OSL `(modules ...)` field and that doesn't support qualified already, so the business gets tricky.

> * Why does `c_sources` need the check? We're not making the C symbols qualified, are we? (they aren't even qualified by library, I think)

I added the check as it does support `unqualified`, thus IMO it should error.

> to make "local" cheap we could just reverse it

Indeed, I didn't complicate the logic as at use point you'd have to reverse it anyways for most cases, so the cost is just delayed. But I always feel pretty bad when tail-appending :)

> I'm happy if this is merged and any follow-up changes happen later.

Cool, I'll merge tonight then I think if no further blocking comments do appear.

Reply:
> have to interpret the OSL (modules ...) field

Oh, I see. But you presumably solve that problem in the Coq PR somehow...

> I added the check as it does support `unqualified`

Ok (although I'm not sure what "support" means for C)

> you'd have to reverse it anyways for most cases, so the cost is just delayed

Only where it's used, right? (which is "nowhere" at the moment :-D)

Reply:
> But you presumably solve that problem in the Coq PR somehow...

In the Coq PR we interpret the OSL for the `(coqlib ...)` stanza where `(modules ...)` refers to coq modules, thus the problem for OCaml is still there.

> Ok (although I'm not sure what "support" means for C)

Good point, I think that what it does is to "flatten" the C sources; that indeed can be useful in some contexts.

> Only where it's used, right? (which is "nowhere" at the moment :-D)

Yup; note that it could be used in more than one place tho so the reversal would be wasteful I think, however the current code is too naive, I've improved it a bit.

Reply:
Oh, I see. Thanks.